game-headed/build.gradle

```diff
@@ -20,6 +18,14 @@ dependencies {
     compile project(':game-core')
 }
 
+install4j {
+    gradle.taskGraph.whenReady {
+        if (gradle.taskGraph.hasTask(generateInstallers)) {
```

Comment:
Correct me if I'm wrong, but this line should always return true, because we are creating the task ourselves. 
So if this line always returns true, there's no reason to wait for the taskGraph to be ready, so both closures could be removed from my understanding.

Reply:
Yeah, this is a little confusing.  `Gradle#getTaskGraph()` returns a [`TaskExecutionGraph`](https://docs.gradle.org/current/javadoc/org/gradle/api/execution/TaskExecutionGraph.html):

> The `TaskExecutionGraph` maintains an execution plan of tasks to be executed (or which have been executed), ...
>
> The `TaskExecutionGraph` is populated only after all the projects in the build have been evaulated. It is empty before then.

That is, a task will only be present in the task execution graph if it has been, or will be, executed--not simply if it exists.

For example, if you run `:game-headed:check`, the `generateInstallers` task _will not_ be present in the task execution graph.  But, if you run `:game-headed:release`, the `generateInstallers` task _will_ be present in the task execution graph (because it is a transitive dependency of the `release` task).

The purpose behind evaluating this condition is so that a developer is only required to define the `install4jHomeDir` project property if they want to run the `:game-headed:generateInstallers` task.  Imagine a new developer who wants to simply run the client via `:game-headed:run` or run `:game-headed:check` before they submit a PR.  Without this check, they will be required to define `install4jHomeDir`.  Since install4j requires a product key, most developers won't want to install it, and then will just have to set the property to some dummy folder, and due to the way the Gradle `file()` function works, that folder must exist or it will raise an exception.  Bottom line, most devs will rarely (if ever) run `:game-headed:generateInstallers` and shouldn't be required to define a project property that will never be used and otherwise has no good default since install4j won't typically be installed.

Another purpose for this condition is to avoid having to install install4j in Travis if the build will not produce a release (i.e. run the `release` task).  The way the Travis buildscript is configured is that the _.travis/install_install4j_ script will only be run for a release build (i.e. builds on the `master` branch).  If the above condition wasn't present, we'd have to execute _.travis/install_install4j_ in the `before_script` phase before the first Gradle invocation so that the `install4jHomeDir` property would be written to _~/.gradle/gradle.properties_.  As that script downloads the install4j installer (34 MiB), it would be a waste of bandwidth if it weren't actually used (e.g. in a PR build).

Reply:
Thanks for the clarification.
Makes sense.

Reply:
I had to think about this a little bit longer:
With this change what's essentially happening is that we only set the install4j directory when we require it.
However what would speak against setting it every time/setting it every time the property is being set if the empty string results in undefined behaviour.

Reply:
> With this change what's essentially happening is that we only set the install4j directory when we require it.

Yes, exactly.

> However what would speak against setting it every time/setting it every time the property is being set if the empty string results in undefined behaviour.

Not sure I understand what you're proposing.  Could you please clarify or add a code snippet for what you're thinking?

Reply:
What I'm saying is that I don't see a difference in behaviour when using this code instead:
```groovy
install4j {
  if (project.hasProperty('install4jHomeDir')) { // or something to check if the directory is set
    installDir = file(install4jHomeDir) 
  }
}
```

Reply:
Yes, that does seem equivalent.  I'll try running a full Travis build later this evening to ensure it works as expected.  I'll also make sure the error message is meaningful when running the `release` target without setting the `install4j.installDir` property.

Reply:
So, using your above proposal, if `install4jHomeDir` is not set, the following error is reported by Gradle:

```
Execution failed for task ':game-headed:generateInstallers'.
> property "installDir" must be set
```

Do you think that's sufficient context (along with the code in the buildscript obviously) for a new dev to quickly realize what the problem is?

The only other alternative I could think of is the following:

```gradle
install4j {
    installDir = file(install4jHomeDir)
}
```

And provide the following _gradle.settings_ file in the project:

```properties
# If you plan on building the installers, you must install install4j and define this
# property on the command line (e.g. -Pinstall4jHomeDir=...) or in
# ~/.gradle/gradle.properties
install4jHomeDir=.
```

That produces the following error message if the user has not overridden `install4jHomeDir` on the command-line or in _~/.gradle/gradle.properties_:

```
Execution failed for task ':game-headed:generateInstallers'.
> Executable "/home/ssoloff/Projects/triplea-game/triplea/game-headed/bin/install4jc" does not exist
```

Reply:
I'm intrigued by your second proposal and would prefer it, but I'd be ok with the first proposal as well.
I'll leave it up to you.

Reply:
#4009